### PR TITLE
CBG-804: OpenID Connect Library Upgrade

### PR DESCRIPTION
**Pull request by sarathkumarsivan**
This PR was moved from https://github.com/couchbase/sync_gateway/pull/4545.

### CODE REVIEW

File: rest/oidc_api_test.go [hunk 1]
```diff
@@ -547,7 +547,7 @@ func TestOpenIDConnectAuth(t *testing.T) {
 			authURL:         "/db/_oidc?provider=foo&offline=true",
 			forceAuthError: forceError{
 				errorType:            callbackTokenExchangeErr,
-				expectedErrorCode:    http.StatusInternalServerError,
+				expectedErrorCode:    http.StatusUnauthorized,
```
Comment (adamcfraser) [inline rest/oidc_api_test.go]: Do these represent a change in behaviour between the previous and current OIDC implementations?  Can you remind me of the rationale for this change?  Same applies to the two scenarios below.

File: rest/oidc_api_test.go [hunk 2]
```diff
@@ -684,7 +684,7 @@ func TestOpenIDConnectAuth(t *testing.T) {
 			if tc.providers["foo"].IncludeAccessToken {
 				assert.Equal(t, wantTokenResponse.AccessToken, receivedToken.AccessToken, "access_token mismatch")
 				assert.Equal(t, wantTokenResponse.TokenType, receivedToken.TokenType, "token_type mismatch")
-				assert.Equal(t, wantTokenResponse.Expires, receivedToken.Expires, "expires_in mismatch")
+				assert.True(t, wantTokenResponse.Expires >= receivedToken.Expires, "expires_in mismatch")
```
Comment (adamcfraser) [inline rest/oidc_api_test.go]: What's causing a non-exact match here and below?

File: rest/oidc_api_test.go [hunk 3]
```diff
@@ -249,7 +249,7 @@ func (s *mockAuthServer) tokenHandler(w http.ResponseWriter, r *http.Request) {
 		AccessToken:  "7d1d234f5fde713a94454f268833adcd39835fe8",
 		RefreshToken: "e08c77351221346153d09ff64c123b24fc4c1905",
 		TokenType:    "Bearer",
-		Expires:      time.Now().Add(5 * time.Minute).UTC().Second(),
+		Expires:      time.Now().Add(60 * time.Minute).UTC().Second(),
```
Comment (adamcfraser) [inline rest/oidc_api_test.go]: Why is this change needed?

File: rest/handler.go [hunk 4]
```diff
@@ -329,8 +329,8 @@ func (h *handler) checkAuth(context *db.DatabaseContext) (err error) {
 		if context.Options.UnsupportedOptions.OidcTestProvider.Enabled && strings.HasSuffix(h.rq.URL.Path, "/_oidc_testing/token") {
 			if username, password := h.getBasicAuth(); username != "" && password != "" {
 				provider := context.Options.OIDCOptions.Providers.GetProviderForIssuer(issuerUrlForDB(h, context.Name), testProviderAudiences)
-				if provider != nil && provider.ClientID != nil && provider.ValidationKey != nil {
-					if *provider.ClientID == username && *provider.ValidationKey == password {
+				if provider != nil && provider.ClientID != "" && provider.ValidationKey != nil {
```
Comment (adamcfraser) [inline rest/handler.go]: Looks like this was just guarding against nil pointer exceptions when comparing ClientID below.  I don't think there's any reason to check for empty string, is there?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Merge issue on manifest, but also some changes in test results that I didn't expect.  Have a look, and if you think they are appropriate, please provide some details explaining why.

File: rest/handler.go (hunk 4 above)
Comment (sarathkumarsivan) [reply rest/handler.go]: That's fair. We're already  checking provider.ClientID == username down below. Fixed in latest revision.

**Review by sarathkumarsivan — COMMENTED** [review]
- Remove gocbconnstr dependency changes committed against this PR from manifest file.
- Remove client ID empty check which is obsolete.
- Return status 500 internal server error when there is any failure in auth/refresh token exchange.
- Return status 401 unauthorized error when no token_id is received from OAuth server in auth/refresh responses.
- Fix the token expiration time sent from fake OAuth server and correct the assertion logic.

**Comment by sarathkumarsivan** [comment]
> Merge issue on manifest, but also some changes in test results that I didn't expect. Have a look, and if you think they are appropriate, please provide some details explaining why.

Thank you Adam, sure I've updated the details.

File: rest/oidc_api_test.go (hunk 1 above)
Comment (adamcfraser) [reply rest/oidc_api_test.go]: Per https://openid.net/specs/openid-connect-core-1_0.html#RefreshErrorResponse, on error it's going to return these errors: https://tools.ietf.org/html/rfc6749#section-5.2.  That looks like we'd be able to differentiate between invalid requests and failed authentications.

We should only be returning a 401 for failed authentications - any unexpected errors should continue to be treated as 500.

Comment (sarathkumarsivan) [reply rest/oidc_api_test.go]: Make sense, thanks Adam for clarifying this. Revised in latest revision.

**Review by adamcfraser — APPROVED** [review]
